### PR TITLE
Refactor ldap filter

### PR DESCRIPTION
**Pull request by arhimondr**
Cache authentication per credentials. In is important to cache authentication,
as we are going to use LDAP authentication for internal communication. Using LDAP
authentication without any caching will lead to severe performance degradation.

### CODE REVIEW

**Comment by arhimondr** [comment]
@anusudarsan Please review

**Comment by arhimondr** [comment]
@anusudarsan @sanjay990 You should better keep the implementation and the product tests on the same branch. In order to test my changes i had to cherry-pick product test first, run them, and than reset.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 1]
```diff
@@ -106,153 +131,171 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
...
+
+    private static void sendChallenge(HttpServletRequest request, HttpServletResponse response, String errorMessage)
+            throws IOException
+    {
+        long contentLength = request.getContentLengthLong();
+        if (contentLength > 0) {
+            skipFully(request.getInputStream(), contentLength);
+        }
+        response.setHeader(HttpHeaders.WWW_AUTHENTICATE, "Basic realm=\"presto\"");
+        response.sendError(UNAUTHORIZED.code(), errorMessage);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: setStatus instead of sendError

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 2]
```diff
@@ -106,153 +131,171 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
...
+                return;
+            }
+            response.sendError(e.getStatus().code(), e.getMessage());
+        }
+    }
+
+    private static void sendChallenge(HttpServletRequest request, HttpServletResponse response, String errorMessage)
+            throws IOException
+    {
+        long contentLength = request.getContentLengthLong();
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: Please extract this to a separate commit

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 3]
```diff
@@ -106,153 +131,171 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
...
+        if (contentLength > 0) {
+            skipFully(request.getInputStream(), contentLength);
+        }
+        response.setHeader(HttpHeaders.WWW_AUTHENTICATE, "Basic realm=\"presto\"");
+        response.sendError(UNAUTHORIZED.code(), errorMessage);
+    }
+
+    private static Credentials getCredentials(String header)
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: I know it's a bit of work, but could you make a separate commit where you extract the getCredentials/getPrincipal logic, before you refactor it? That would make it clearer what has changed.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 4]
```diff
@@ -106,153 +131,171 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
...
+            throws AuthenticationException
+    {
+        if (header == null) {
+            throw new AuthenticationException(UNAUTHORIZED, "Unauthorized");
+        }
+        if (!header.startsWith(BASIC_AUTHENTICATION_PREFIX)) {
+            throw new AuthenticationException(BAD_REQUEST, "Basic authentication is expected");
+        }
+        String base64Credentials = header.substring(BASIC_AUTHENTICATION_PREFIX.length());
+        String credentials = new String(Base64.getDecoder().decode(base64Credentials), UTF_8);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: This version of the code handles an incorrect request differently than before: now, if there were more than two \s-delimited parts, this decode() method will throw an exception and you'll get 500, rather than an "Invalid credentials" error (see line 118 of the old code).

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 5]
```diff
@@ -106,153 +131,171 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
...
             }, servletResponse);
         }
         catch (AuthenticationException e) {
-            // log the failure since the exception goes directly to the client
+            log.debug(e, "LDAP authentication failed");
+            if (e.getStatus() == UNAUTHORIZED) {
+                sendChallenge(request, response, e.getMessage());
+                return;
+            }
+            response.sendError(e.getStatus().code(), e.getMessage());
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: The original code sends a challenge for SC_BAD_REQUEST as well, rather than an error.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 6]
```diff
@@ -106,153 +131,171 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
...
-        String user = credentials[0];
-
         try {
-            String password = credentials[1];
-            if (user.isEmpty() || password.isEmpty()) {
-                throw new AuthenticationException("Username or password is empty");
-            }
-
-            Hashtable<String, String> environment = getBasicEnvironment();
-            String principal = genericLdapBinder.getUserBindSearchPattern().replaceAll("\\$\\{USER\\}", user);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java:143]: This functionality is not present in the new implementation

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 7]
```diff
@@ -106,153 +131,171 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
...
+            throws AuthenticationException
+    {
+        Map<String, String> environment = createEnvironment(credentials);
+        String user = credentials.getUser();
+        InitialDirContext context = null;
+        try {
+            context = createDirContext(environment);
+            checkForGroupMembership(user, context);
+            log.debug("Authentication successful for user %s.", user);
+            return new LdapPrincipal(user);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: Why use this wrapper class? Does Principal not let you just pass a user in?

**Review by cawallin — COMMENTED** [review]
I haven't finished reviewing (still need to get to group stuff), I'll do that tomorrow.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 6 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java:143]: It is there. I extracted it to the separate method  `private String createPrincipal(String user)`

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 5 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: Actually i was wrong about those method. The only difference between `setStatus` and `sendMessage` is that the latest accepts `message` as a parameter. So, as long as we won't to return specific message with the status it is fine to use this method. It won't produce any stacktrace it the response body.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 1 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: ditto about `sendError`

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 3 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: I think those commits should be squashed anyway. I don't think that Facebook is going to merge the caches as a local field.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 7 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: I just think that such interface is better. Like `authenticate` method accepts `Credentials` and returns `Principal` if the authentication is successful. I believe it is more clear than just return `String`.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 4 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: It throws `AuthenticationException` that is further processed as a special case, returning the status stored in this exception

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: But i found another issue. If you put something dummy instead of base64 encoded data, it will throw `500` instead of BadRequest. Gonna fix that.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: Oh. I understand now your first comment.

Ok. This is yet another bug i had to fix. Due to the Basic authentication specification only the user name can't contain the `:` character. Password can. And the credentials should be split only by the first `:`. So then the password can contain colon.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 6 above)
Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java:143]: Oh sorry, I saw that but forgot to remove the comment

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 3 above)
Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: Fair point

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 5 above)
Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: Do you mean `sendError` instead of `sendMessage` in the above explanation?

Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: Though also,  sendChallenge does a few other things:
- Iterate through the whole request
- response.setHeader(HttpHeaders.WWW_AUTHENTICATE, "Basic realm=\"presto\"");

Do we need to do those things for SC_BAD_REQUEST, or is sending the error sufficient?

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 8]
```diff
@@ -106,153 +130,178 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
...
+        }
+
         checkState(serverConfig.getUserBaseDistinguishedName() != null, "Base distinguished name (DN) for user %s is null", user);
 
-        try {
-            String searchFilter = searchFilterPattern.replaceAll("\\$\\{USER\\}", user);
-            GroupAuthorizationFilter groupFilter = new GroupAuthorizationFilter(serverConfig.getUserBaseDistinguishedName(), searchFilter);
+        String searchFilter = pattern.get().replaceAll("\\$\\{USER\\}", user);
+        String userBaseDistinguishedName = serverConfig.getUserBaseDistinguishedName();
+        String groupName = firstNonNull(serverConfig.getLdapServerType().equals(GENERIC) ? serverConfig.getGroupDistinguishedName() : "", "");
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: This logic is a) pretty confusing and b) different than what was used before:
`String groupNameIfPresent = firstNonNull(serverConfig.getLdapServerType().equals(GENERIC) ? null : serverConfig.getGroupDistinguishedName(), "");`
What is it trying to do? It may be worthwhile to extract a method here and actually write out the if statements rather than using a ternary + firstNonNull

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 9]
```diff
@@ -106,153 +130,178 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
...
+            return;
+        }
+
         checkState(serverConfig.getUserBaseDistinguishedName() != null, "Base distinguished name (DN) for user %s is null", user);
 
-        try {
-            String searchFilter = searchFilterPattern.replaceAll("\\$\\{USER\\}", user);
-            GroupAuthorizationFilter groupFilter = new GroupAuthorizationFilter(serverConfig.getUserBaseDistinguishedName(), searchFilter);
+        String searchFilter = pattern.get().replaceAll("\\$\\{USER\\}", user);
+        String userBaseDistinguishedName = serverConfig.getUserBaseDistinguishedName();
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: requireNonNull

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 10]
```diff
@@ -106,153 +130,178 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
...
-                log.debug("Authorization failed for user. " + errorMessage);
-                throw new RuntimeException("Unauthorized user: " + errorMessage);
             }
-            log.debug("Authorization succeeded for user %s in group %s", user, groupNameIfPresent);
         }
-        catch (ExecutionException e) {
-            throw Throwables.propagate(e);
+
+        if (!authorized) {
+            String message = format("Unauthorized user: User %s not a member of the group %s", user, groupName);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: A little tangential to your changes, but can the group filter ever fail if the groupName is ""? It seems weird to be doing the LDAP group search if there's no group name specified.

**Review by cawallin — COMMENTED** [review]
I'd like a bit of clarification on the group handling and also on the responses we're returning when we can't parse something.

**Comment by cawallin** [comment]
Also, according to http://docs.oracle.com/javase/tutorial/jndi/ldap/exceptions.html, some of the subclasses of NamingException may indicate somewhat unauthorized-like responses (e.g. AuthenticationNotSupportedException if strong authentication is required -- I know that Kerberos throws unauthorized if you don't have the correct strength keys). Do we want to handle anything for that, or just leave it?

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 9 above)
Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: Actually, this should be annotated as `@NonNull` or something like that in serverConfig. Can you make sure it is?

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 5 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: There is no method called `sendMessage`. 

There are only 2 methods that can set status to the response:
- `setStatus` - sets the status code, and the default message for this code
- `sendError` - allows you to set the status, and the custom status message 

> Iterate through the whole request

No, it is not needed for any other response than `UNAUTHORIZED`. Gonna change that.

> response.setHeader(HttpHeaders.WWW_AUTHENTICATE, "Basic realm=\"presto\"");

I set this header only for `UNAUTHORIZED`

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java (hunk 10 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: I don't understand this code actually. As i understand the group name is set for `ActiveDirectory`, and it is not for `Generic` or `OpenLDAP`. Actually we need to unify everything and leave only  the `GenericLdap`. Because the only difference between them - is the format of base distinguished name and group. I'm going to figure out it in terms of that task.

**Review by cawallin — APPROVED** [review]
Looks good % small comments (requireNonNull and the group handling)
